build_daemon/lib/client.dart

```diff
@@ -0,0 +1,129 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:web_socket_channel/io.dart';
+
+import 'constants.dart';
+import 'data/build_options_request.dart';
+import 'data/build_request.dart';
+import 'data/build_status.dart';
+import 'data/build_target_request.dart';
+import 'data/log_to_paths_request.dart';
+import 'data/serializers.dart';
+import 'data/server_log.dart';
+
+class BuildDaemonClient {
+  IOWebSocketChannel _channel;
+
+  final _buildResultsStreamController =
+      StreamController<BuildResults>.broadcast();
+
+  final _serverLogStreamController = StreamController<ServerLog>.broadcast();
+
+  BuildDaemonClient._();
+
+  Stream<BuildResults> get buildResultsStream =>
+      _buildResultsStreamController.stream;
+  Future<void> get finished async {
+    if (_channel != null) await _channel.sink.done;
+  }
+
+  Stream<ServerLog> get serverLogs => _serverLogStreamController.stream;
+
+  /// Adds a list of build options to be used for all builds.
+  ///
+  /// When this client disconnects these options will no longer be used by the
+  /// build daemon.
+  void addBuildOptions(Iterable<String> options) {
+    var request = BuildOptionsRequest((b) => b..options.replace(options));
+    _channel.sink.add(jsonEncode(serializers.serialize(request)));
+  }
+
+  /// Adds paths to write usage logs to.
+  ///
+  /// When the client disconnects these files will no longer be logged to.
+  void logToPaths(Iterable<String> paths) {
+    var request = LogToPathsRequest((b) => b..paths.replace(paths));
+    _channel.sink.add(jsonEncode(serializers.serialize(request)));
+  }
+
+  /// Registers a build target to be built upon any file change.
+  ///
+  /// Changes that match the patterns in [blackListPattern] will be ignored.
+  ///
+  /// Note that you should provide the "fully qualified" target name. For
+  /// example one should provide //some-target:test_run instead of
```

Comment:
[nit] This is a detail of the Dart test rules on bazel - should it be included here?

Reply:
Removed.